parachain/pallets/ethereum-beacon-client/src/lib.rs

```diff
@@ -252,11 +274,32 @@ pub mod pallet {
 		}
 
 		pub(crate) fn process_update(update: &Update) -> DispatchResult {
+			Self::cross_check_execution_state()?;
 			Self::verify_update(update)?;
 			Self::apply_update(update)?;
 			Ok(())
 		}
 
+		// Cross check to make sure ExecutionHeader not fall behind FinalizedHeader too much, if
+		// that happens just return error so to pause processing FinalizedHeader until
+		// ExecutionHeader catch up
+		fn cross_check_execution_state() -> DispatchResult {
+			let latest_finalized_state: CompactBeaconState =
+				match Self::finalized_beacon_state(Self::latest_finalized_block_root()) {
+					Some(finalized_beacon_state) => finalized_beacon_state,
+					None => return Err(Error::<T>::NotBootstrapped.into()),
+				};
+			let latest_execution_state: ExecutionHeaderState = Self::latest_execution_header();
+			let max_latency = config::EPOCHS_PER_SYNC_COMMITTEE_PERIOD * config::SLOTS_PER_EPOCH;
+			ensure!(
+				latest_execution_state.beacon_slot == 0 ||
+					latest_finalized_state.slot <
+						latest_execution_state.beacon_slot + max_latency as u64,
+				Error::<T>::ExecutionHeaderFallBehindTooMuch
+			);
```

Comment:
With this constraint added we're safe always use `LatestFinalizedState` as the only checkpoint and pruing `FinalizedState`.